### ISSUE TITLE
Bug(s) in paramter_learn

### ISSUE THREAD

**Issue by khosravipasha** [Tractables/DiscriminativeCircuits.jl#13]
1. If paramter_learning is called on `LogisticCircuit` nothing gets updated
2. If paramter_learning is called on `ParamBitCircuit` made from LogisticCircuit, then after one epoch the predictions become all wrong and accuracy drop to `0.09`.

```julia
using Revise
using LogicCircuits
using ProbabilisticCircuits
using DiscriminativeCircuits
using MLDatasets
using DataFrames: DataFrame
using Statistics

function mnist()
    # transposing makes slicing by variable much much faster
    # need to take a copy to physically move the data around
    train_x = collect(Float32, transpose(reshape(MNIST.traintensor(), 28*28, :)))
    test_x  = collect(Float32, transpose(reshape(MNIST.testtensor(), 28*28, :)))
    
    train_x = DataFrame(train_x, :auto)
    test_x = DataFrame(test_x, :auto)
    
    train_y::Vector{Int32} = MNIST.trainlabels()
    test_y::Vector{Int32}  = MNIST.testlabels()
    
    return train_x, train_y, test_x, test_y
end


function main()
    circuit_name = "mnist.circuit"
    lc = zoo_dc(circuit_name);
    println("Loaded zoo_dc($(circuit_name)) with $(num_classes(lc)) classes")
    @time begin
        mnist_train, train_y, mnist_test, test_y = mnist();
        mnist_train, _, mnist_test = threshold(mnist_train[:, 1:end-1], nothing, mnist_test[:, 1:end-1], 0.5);
    end

    # WARMup
    predict_class(lc, mnist_train[1:3,:])
    predict_class(lc, to_gpu(mnist_train[1:3,:]))

    pbc = ParamBitCircuit(lc, mnist_train)

    print_acc() = begin
        @time pred = predict_class(pbc, mnist_test)
        pred .-= 1 # does not return lables its return the index of the label
        acc = mean( pred .== test_y )
        println("\t [CPU] Accuray = $( acc )")

        @time pred_gpu = predict_class(pbc, to_gpu(mnist_test))
        pred_gpu .-=1
        acc = mean( to_cpu(pred_gpu) .== test_y )
        println("\t [GPU] Accuracy = $( acc )")

        println("CPU - GPU Match $( mean( to_cpu(pred_gpu) .== pred ) );")

        pred, pred_gpu
    end

    pred, pred_gpu = print_acc()

    @time begin
        println("[CPU] learn_parameters")
        step_size = 0.1
        num_epochs = 1
        learn_parameters(pbc, mnist_train, train_y; num_epochs, step_size)
    end

    pred, pred_gpu = print_acc()

    @time begin
        println("[GPU] learn_parameters")
        step_size = 0.1
        num_epochs = 1
        mnist_train_gpu = to_gpu(mnist_train)
        mnist_train_y_gpu = to_gpu(train_y)
        learn_parameters(pbc, mnist_train_gpu, mnist_train_y_gpu; num_epochs, step_size)
    end

    pred, pred_gpu = print_acc()

    println(test_y[1:10])
    println(pred[1:10])
    println(to_cpu(pred_gpu)[1:10])

    println(mean(pred .== 0))

    pbc
end
```

```
Loaded zoo_dc(mnist.circuit) with 10 classes
  1.611870 seconds (27.10 k allocations: 921.044 MiB, 4.88% gc time)
  0.233324 seconds (10.39 k allocations: 28.905 MiB)
         [CPU] Accuray = 0.9651
  0.021951 seconds (55.26 k allocations: 2.820 MiB)
         [GPU] Accuracy = 0.9651
CPU - GPU Match 1.0;
[CPU] learn_parameters
  4.471037 seconds (27.11 k allocations: 334.353 MiB, 0.52% gc time)
  0.224069 seconds (10.39 k allocations: 28.905 MiB)
         [CPU] Accuray = 0.098
  0.022307 seconds (55.26 k allocations: 2.820 MiB)
         [GPU] Accuracy = 0.098
CPU - GPU Match 1.0;
[GPU] learn_parameters
  0.073387 seconds (103.69 k allocations: 5.283 MiB)
  0.224260 seconds (10.39 k allocations: 28.905 MiB)
         [CPU] Accuray = 0.098
  0.022001 seconds (55.26 k allocations: 2.820 MiB)
         [GPU] Accuracy = 0.098
CPU - GPU Match 1.0;
Int32[7, 2, 1, 0, 4, 1, 4, 9, 5, 9]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
1.0
```
